Function pointers prototype comment followup

333fred opened the pull request:
Follow up on a number of prototype comments. Again, I recommend reviewing commit-by-commit for this PR.

333fred commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -4051,6 +4078,410 @@ .maxstack  3
+using System;
+unsafe class C
+{
+    static (bool, bool, bool, bool, bool, bool) DoCompare(delegate*<void> func_1a, delegate*<string, void> func_1b)
```
I didn't think it was necessary or useful to verify the exact IL for each type of binary operator conversion, so I split out the operators for just the function pointer to function pointer operators and verified this one function.

333fred commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:
```diff
@@ -323,9 +323,9 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
             ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
             Debug.Assert(funcPtrBuilder.Count == 1);
+            Debug.Assert(funcPtrBuilder[0].Arity == 0);
```
The only time `inferWithDynamic` is used is when `Arity` is `> 0`, so I've added an assert that that's never true (which it can't be, as function pointers cannot define their own type parameters).

333fred commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -71,7 +71,6 @@ internal static class ParameterHelpers
                 diagnostics,
                 allowRefOrOut: true,
                 allowThis: false,
-                // PROTOTYPE(func-ptr): Investigate if cases where this can be false for regular parameters applies to func pointers as well
```
InAttribute cannot avoid being emitted for function pointer types, as it's part of the signature. Existing places that avoid it are non-virtual methods, which can also emit metadata on the parameter for this case. Function pointers have no other location to encode this, so we must do it here.

AlekseyTs replied:
> InAttribute cannot avoid being emitted for function pointer types, as it's part of the signature. Existing places that avoid it are non-virtual methods, which can also emit metadata on the parameter for this case. Function pointers have no other location to encode this, so we must do it here.

Consider capturing this in a comment, so that the next person wouldn't have to figure this out again

---
In reply to: [408483652](https://github.com/dotnet/roslyn/pull/43355#discussion_r408483652) [](ancestors = 408483652)

333fred commented on src/Compilers/CSharp/Test/Semantic/Semantics/UnsafeTests.cs:
```diff
@@ -4306,13 +4306,9 @@ void M()
             var type = typeInfo.Type;
             var conv = model.GetConversion(syntax);
-            Assert.NotNull(type);
+            Assert.Null(type);
```
I was originally thinking we might want to try and replicate today's behavior here, but I think we'd rather be consistent with the behavior of method groups. Today, this is not meaningful, but with function pointers it is.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs this is ready for review. Again, I recommend commit-by-commit for this PR. There are fewer commits, but they are slightly bigger than the previous PR. There are still prototype comments left in the branch after this.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for review.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1918,5 +1918,34 @@ void M(delegate*<void> ptr)
+{
+    static void M(delegate*<string, int, void> ptr1, delegate*<__arglist, void> ptr2)
+    {
+        ptr1(__arglist(string.Empty, 1), 1);
```
>ptr1(__arglist(string.Empty, 1), 1); [](start = 8, length = 36)

Consider testing when __arglist is the last in the argument list, including the case when it is after all parameters that method can take.

AlekseyTs commented:
Is there a specification for supporting ```out``` parameters in function pointers?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/SymbolFactory.vb:
```diff
@@ -56,6 +56,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
         End Function
 
+        Friend Overrides Function IsAcceptedOutAttributeModifierType(type As TypeSymbol) As Boolean
+            ' VB doesn't deal with ref-readonly parameters, function pointers, or return-types.
```
>or return-types. [](start = 79, length = 16)

Is this accurate?

333fred replied:
I just copied the comment from the method above and added function pointers. I'll modify both to specify ref-return types, which is what I suspect it should have been in the first place.

---
In reply to: [409785521](https://github.com/dotnet/roslyn/pull/43355#discussion_r409785521) [](ancestors = 409785521)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -305,11 +305,11 @@ static ImmutableArray<FunctionPointerParameterSymbol> makeParametersFromMetadata
                 return param.IsByRef switch
                 {
                     false => RefKind.None,
                     true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => hasInRefKind,
```
>true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => hasInRefKind, [](start = 20, length = 89)

How do we deal with a situation when there are both modifiers? Please add a test for a situation like that.

333fred replied:
This is a good test, will add.

---
In reply to: [409787440](https://github.com/dotnet/roslyn/pull/43355#discussion_r409787440) [](ancestors = 409787440)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -305,11 +305,11 @@ static ImmutableArray<FunctionPointerParameterSymbol> makeParametersFromMetadata
                 {
                     false => RefKind.None,
                     true when CustomModifierUtils.HasInAttributeModifier(paramRefCustomMods) => hasInRefKind,
+                    true when CustomModifierUtils.HasOutAttributeModifier(paramRefCustomMods) => RefKind.Out,
```
> true when CustomModifierUtils.HasOutAttributeModifier(paramRefCustomMods) => RefKind.Out, [](start = 19, length = 90)

This doesn't look like the right thing to do for a return RefKind. Please add a test

333fred replied:
I'll add a test, but I'll also add an assert. The return parameter is decoded only allowing `InAttribute`, so if it encountered an out during decode we'd throw and never construct a function pointer type symbol at all.

---
In reply to: [409789040](https://github.com/dotnet/roslyn/pull/43355#discussion_r409789040) [](ancestors = 409789040)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -91,6 +92,28 @@ static void symbolValidator(ModuleSymbol module)
+            var comp = CompileAndVerifyFunctionPointers(@"
+class C
+{
+    public unsafe void M(delegate*<out C, out string, ref int[]> param1) => throw null;
```
>ref [](start = 54, length = 3)

Do we have a test for an attempt to use ```out``` on a return type?

333fred replied:
Added to `FunctionPointerTypeSymbolTests.InvalidReturnModifiers`

---
In reply to: [409814570](https://github.com/dotnet/roslyn/pull/43355#discussion_r409814570) [](ancestors = 409814570)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -2610,35 +2633,96 @@ .locals init (delegate*<string,object> V_0)
+        [Theory]
+        [InlineData("in")]
+        [InlineData("ref")]
+        public void AddressOf_Initializer_Overloads(string refType)
```
>AddressOf_Initializer_Overloads [](start = 20, length = 31)

Do we have an error test for ref/out mismatch?

333fred replied:
This was supposed to be `FunctionPointerToFunctionPointerInvalid_ParameterVariance`

---
In reply to: [409817853](https://github.com/dotnet/roslyn/pull/43355#discussion_r409817853) [](ancestors = 409817853)

333fred replied:
But I'll add a method group resolution test as well.

---
In reply to: [410493966](https://github.com/dotnet/roslyn/pull/43355#discussion_r410493966) [](ancestors = 410493966,409817853)

333fred replied:
Additional cases added to `AddressOf_RefsMustMatch`.

---
In reply to: [410494067](https://github.com/dotnet/roslyn/pull/43355#discussion_r410494067) [](ancestors = 410494067,410493966,409817853)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -4271,13 +4575,33 @@ private static void VerifyFunctionPointerSymbol(TypeSymbol type, CallingConventi
+                    Assert.True(false, $"Cannot have a return ref kind of {signature.RefKind}");
+                    break;
+
+                case RefKind.RefReadOnly:
```
>case RefKind.RefReadOnly: [](start = 16, length = 25)

What are other possible values? Please assert expectations around modifiers for them too.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -4271,13 +4575,33 @@ private static void VerifyFunctionPointerSymbol(TypeSymbol type, CallingConventi
             {
                 Assert.Equal(argumentVerifiers[i].RefKind, signature.Parameters[i].RefKind);
                 argumentVerifiers[i].TypeVerifier(signature.Parameters[i].Type);
+                switch (signature.RefKind)
```
>switch (signature.RefKind) [](start = 16, length = 26)

What are we testing here? How is it related to parameters?

333fred replied:
We're ensuring that parameters marked as `in` or `out` have the appropriate metadata.

---
In reply to: [409819988](https://github.com/dotnet/roslyn/pull/43355#discussion_r409819988) [](ancestors = 409819988)

AlekseyTs replied:
> We're ensuring that parameters marked as in or out have the appropriate metadata.

Is this really what is happening? What are signature.RefKind, signature.RefCustomModifiers, signature.RefCustomModifiers? How are they related to parameters?

---
In reply to: [409838517](https://github.com/dotnet/roslyn/pull/43355#discussion_r409838517) [](ancestors = 409838517,409819988)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -618,11 +622,16 @@ public void FunctionPointerToFunctionPointerInvalid_ParameterVariance()
 unsafe class C
 {
-    void M(delegate*<ref object, void> param1)
+    void M(delegate*<ref object, void> param1, delegate*<ref object, void> param2)
```
>delegate*<ref object, void> [](start = 47, length = 27)

This is the same type as for param1. Why is it useful to add it?

333fred replied:
Whoops, this should have been an out. Will fix.

---
In reply to: [409822981](https://github.com/dotnet/roslyn/pull/43355#discussion_r409822981) [](ancestors = 409822981)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -91,6 +92,28 @@ static void symbolValidator(ModuleSymbol module)
+                var m = c.GetMethod("M");
+                var funcPtr = m.ParameterTypesWithAnnotations[0].Type;
+
+                VerifyFunctionPointerSymbol(funcPtr, CallingConvention.Default,
```
>VerifyFunctionPointerSymbol [](start = 16, length = 27)

We should also validate custom modifiers for source symbols.

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/Members.cs:
```diff
@@ -22,7 +22,6 @@ namespace Microsoft.Cci
     /// <summary>
     /// Specifies how the caller passes parameters to the callee and who cleans up the stack.
     /// </summary>
-    // PROTOTYPE(func-ptr): represent `unmanaged`
```
>// PROTOTYPE(func-ptr): represent `unmanaged` [](start = 4, length = 45)

Is this what ```Standard``` already means?

333fred replied:
No, `Standard` is the `StdCall` convention. `unamanged` is supposed to represent whatever the platform-default convention is. `StdCall` is an unfortunate name for the convention, but it's also what everyone calls it so we're stuck with it.

---
In reply to: [409826087](https://github.com/dotnet/roslyn/pull/43355#discussion_r409826087) [](ancestors = 409826087)

AlekseyTs replied:
> No, Standard is the StdCall convention. unamanged is supposed to represent whatever the platform-default convention is. StdCall is an unfortunate name for the convention, but it's also what everyone calls it so we're stuck with it.

I used Standard in term of this enum. 
So, is ```unmanaged``` already represented in this enum or not?

---
In reply to: [409839891](https://github.com/dotnet/roslyn/pull/43355#discussion_r409839891) [](ancestors = 409839891,409826087)

333fred replied:
No, the addition is tracked in [https://github.com/dotnet/roslyn/issues/39865](https://github.com/dotnet/roslyn/issues/39865).

---
In reply to: [409850766](https://github.com/dotnet/roslyn/pull/43355#discussion_r409850766) [](ancestors = 409850766,409839891,409826087)

AlekseyTs commented:
>         var i1 = &M;

Is this an error scenario? Please assert diagnostics to eliminate all the guess work.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/UnsafeTests.cs:4290 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

AlekseyTs commented:
Done with review pass (iteration 7)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1918,5 +1918,34 @@ void M(delegate*<void> ptr)
         }
+
+        [Fact]
+        public void ArglistCannotBeUsedWithFunctionPointers()
```
>ArglistCannotBeUsedWithFunctionPointers [](start = 20, length = 39)

It looks like it is possible to specify VARARG in metadata. Please add a test for consuming metadata like that.

333fred replied:
Added to `CodeGenFunctionPointersTests.UnsupportedCallingConventionInMetadata`

---
In reply to: [409838193](https://github.com/dotnet/roslyn/pull/43355#discussion_r409838193) [](ancestors = 409838193)

AlekseyTs commented:
>             | SignatureCallingConvention.ThisCall

Are we able to properly handle this calling convention at the call site? From the comment above: "the this pointer is pushed on the stack last". Are we doing something like that?

---
Refers to: src/Compilers/Core/Portable/PEWriter/Members.cs:81 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

333fred commented:
>             | SignatureCallingConvention.ThisCall

I don't believe that's our responsibility, this should be the JIT's job. I'll confirm that with the runtime team though.

---
In reply to: [614893461](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614893461) [](ancestors = 614893461)

---
Refers to: src/Compilers/Core/Portable/PEWriter/Members.cs:81 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

333fred commented:
No, I'll put it on the list for LDM.

---
In reply to: [614836583](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614836583) [](ancestors = 614836583)

333fred commented:
Proposal is in PR: [https://github.com/dotnet/csharplang/pull/3368](https://github.com/dotnet/csharplang/pull/3368)

---
In reply to: [614970953](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614970953) [](ancestors = 614970953,614836583)

cston commented on src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs:
```diff
@@ -655,7 +654,7 @@ private void GetEnumOperations(BinaryOperatorKind kind, BoundExpression left, Bo
+            if ((object)leftType != null || (object)rightType != null || left.Type is FunctionPointerTypeSymbol || right.Type is FunctionPointerTypeSymbol)
             {
                 // The pointer comparison operators are all "void* OP void*".
                 GetPointerComparisonOperators(kind, results);
```
>GetPointerComparisonOperators [](start = 16, length = 29)

Do we support `<` and `>` for function pointers but not pointer arithmetic?

333fred replied:
Unfortunately, yes. I was considering trying to not support `>` and `<`, but they're defined for `void*` so I don't have an option there. But the arithmetic operators are defined for specific pointer types except `void*`, so we can say "these are not defined for function pointer types".

---
In reply to: [410505291](https://github.com/dotnet/roslyn/pull/43355#discussion_r410505291) [](ancestors = 410505291)

333fred commented:
>             | SignatureCallingConvention.ThisCall

Yeah, we don't have anything special to do here.

---
In reply to: [614942213](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614942213) [](ancestors = 614942213,614893461)

---
Refers to: src/Compilers/Core/Portable/PEWriter/Members.cs:81 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

333fred commented:
>         private const SignatureCallingConvention SignatureCallingConventionMask =

`CodeGenFunctionPointersTests.CallingConventions` has round-trip tests for everything but fastcall and varargs.

---
In reply to: [614894682](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614894682) [](ancestors = 614894682)

---
Refers to: src/Compilers/Core/Portable/PEWriter/Members.cs:77 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

333fred commented:
>         private const SignatureCallingConvention SignatureCallingConventionMask =

I'm going to throw on both of these for now, and add a test. The updated calling convention proposal will allow FastCall, and we can determine if we want language support for varargs at that point.

---
In reply to: [616805740](https://github.com/dotnet/roslyn/pull/43355#issuecomment-616805740) [](ancestors = 616805740,614894682)

---
Refers to: src/Compilers/Core/Portable/PEWriter/Members.cs:77 in 3686c83. [](commit_id = 3686c83128a82aa4f7abb3db3a6c5464c64375cf, deletion_comment = False)

jcouv commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -725,6 +725,9 @@ private TypeSymbol GetTypeOfTypeDef(TypeDefinitionHandle typeDef, out bool isNoP
 
                     switch (allowedRequiredModifierType)
                     {
+                        case AllowedRequiredModifierType.System_Runtime_InteropServices_InOrOutAttribute:
```
@333fred Aleksey spotted that we're modifying the same code in different feature branches. Would you be able to make the same change to MetadataDecoder.cs as [here](https://github.com/dotnet/roslyn/pull/43275/files#diff-2231239cabf144e639ea18f4f0ef8627) so that we don't need to worry about merging later on?

For init-only/records, `DecodeModifiersOrThrow` needs to output `out AllowedRequiredModifierType requiredModifiersFound` instead of a `bool`, and make `AllowedRequiredModifierType` be flags so they can be combined. The few callers that consume the output are easily updated.

Let me know it's that would be okay. We could also just wait and merge when the time comes (your changes to this file don't look too bad, so I'm not very worried). Thanks

333fred replied:
I can take a look at this later. I'm trying to get this in today or tomorrow and I'd prefer to keep these changes isolated.

---
In reply to: [411879270](https://github.com/dotnet/roslyn/pull/43355#discussion_r411879270) [](ancestors = 411879270)

333fred replied:
I've unified this.

---
In reply to: [412402730](https://github.com/dotnet/roslyn/pull/43355#discussion_r412402730) [](ancestors = 412402730,411879270)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [614879209](https://github.com/dotnet/roslyn/pull/43355#issuecomment-614879209) [](ancestors = 614879209)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -1187,16 +1190,43 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
+                {
+                    bool hasOut = false;
+                    bool hasIn = false;
+                    foreach (var modifier in info.CustomModifiers)
```
>foreach (var modifier in info.CustomModifiers) [](start = 20, length = 46)

I think this is where approach from [https://github.com/dotnet/roslyn/pull/43275](https://github.com/dotnet/roslyn/pull/43275) would work much better. We would be able to perform the check without having to iterate through modifiers and doing type checks again.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitAddress.cs:
```diff
@@ -120,6 +120,18 @@ private LocalDefinition EmitAddress(BoundExpression expression, AddressKind addr
 
                     goto default;
 
+                case BoundKind.FunctionPointerInvocation:
```
>case BoundKind.FunctionPointerInvocation: [](start = 16, length = 41)

Consider checking other places where we deal with ```BoundKind.Call``` in the ```CodeGenerator``` and making sure we handle ```BoundKind.FunctionPointerInvocation```, if need to.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -365,7 +366,15 @@ internal int GetHashCodeNoParameters()
+            if (CallingConvention.IsCallingConvention(CallingConvention.ExtraArguments) ||
+                CallingConvention.IsCallingConvention(CallingConvention.FastCall))
+            {
+                info = new CSDiagnosticInfo(ErrorCode.ERR_UnsupportedCallingConvention, ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat));
```
>info = new CSDiagnosticInfo(ErrorCode.ERR_UnsupportedCallingConvention, ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat)); [](start = 16, length = 135)

We probably still want to do the ```else``` part and then merge. I believe some errors are higher in priority than others.

AlekseyTs replied:
>ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat) [](start = 88, length = 61)

This looks like a premature string allocation, we do not format symbols until absolutely necessary. If default formatting doesn't work, we have a FormattedSymbol helper type, or something like that.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1918,5 +1942,42 @@ void M(delegate*<void> ptr)
+                //         ptr1(null, __arglist(string.Empty, 1));
+                Diagnostic(ErrorCode.ERR_BadArgType, "__arglist(string.Empty, 1)").WithArguments("2", "__arglist", "int").WithLocation(7, 20),
+                // (8,9): error CS8756: Function pointer 'delegate*<string,int,void>' does not take 3 arguments
+                //         ptr1(null, 1, __arglist(string.Empty, 1));
```
>ptr1(null, 1, __arglist(string.Empty, 1)); [](start = 27, length = 42)

Do we have a test for this scenario when pointer coming from metadata is VARARG?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1918,5 +1942,42 @@ void M(delegate*<void> ptr)
+
+            comp.VerifyDiagnostics(
+                // (4,64): error CS1031: Type expected
+                //     static void M(delegate*<string, int, void> ptr1, delegate*<__arglist, void> ptr2)
```
>__arglist [](start = 82, length = 9)

It would be good to assert value of MethodSymbol.IsVararg for this function pointer signature.

AlekseyTs replied:
Also how many parameters we have and what the types are.

---
In reply to: [413314469](https://github.com/dotnet/roslyn/pull/43355#discussion_r413314469) [](ancestors = 413314469)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -51,12 +57,14 @@ private CSharpCompilation CreateCompilationWithFunctionPointers(string source, I
     public unsafe delegate* {conventionString}<string, int> M() => throw null;
 }}", symbolValidator: symbolValidator);
 
+            symbolValidator((ModuleSymbol)((Symbols.PublicModel.ModuleSymbol)verifier.Compilation.SourceModule).UnderlyingSymbol);
```
>(ModuleSymbol)((Symbols.PublicModel.ModuleSymbol)verifier.Compilation.SourceModule).UnderlyingSymbol [](start = 28, length = 100)

```((CSharpCompilation)verifier.Compilation).SourceModule```?

AlekseyTs replied:
And below as well.

---
In reply to: [413317301](https://github.com/dotnet/roslyn/pull/43355#discussion_r413317301) [](ancestors = 413317301)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -56,7 +56,8 @@ class C
         delegate*<params string> p6,
-        delegate*<ref ref string> p7)
+        delegate*<ref ref string> p7,
+        delegate*<out string> p8)
```
>delegate*<out string> [](start = 8, length = 21)

It would be good to assert RefKind for this symbol.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -387,6 +875,7 @@ public void UnsupportedCallingConventionInMetadata()
 .class public auto ansi beforefieldinit C
        extends [mscorlib]System.Object
 {
+    .field public method vararg void*() 'Field'
```
>method vararg void*() [](start = 18, length = 21)

It would be good to assert value of MethodSymbol.IsVararg for this function pointer signature.

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -6082,4 +6082,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>The type of a local declared in a fixed statement cannot be a function pointer type.</value>
   </data>
+  <data name="ERR_UnsupportedCallingConvention" xml:space="preserve">
+    <value>The calling convention of '{0}' is not supported by C#.</value>
```
>by C# [](start = 60, length = 5)

It doesn't look like we ever add "by C#" to error messages.

AlekseyTs replied:
We often stop at "not supported" or say "not supported by the language"

---
In reply to: [413328055](https://github.com/dotnet/roslyn/pull/43355#discussion_r413328055) [](ancestors = 413328055)

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs:
```diff
@@ -80,7 +79,15 @@ internal static class ParameterHelpers
                 {
-                    ImmutableArray<CustomModifier> customModifiers = ConditionallyCreateInModifiers(refKind, addRefReadOnlyModifier, binder, diagnostics, syntax);
+                    // Non-function pointer locations have other locations to encode in/ref readonly/outness. For function pointers,
+                    // these modreqs are the only locations that this can be encoded. If that changes, we should update this.
```
>that [](start = 60, length = 4)

"where"?

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -145,15 +184,456 @@ static void symbolValidator(ModuleSymbol module)
+    }
+}";
+
+            //CreateCompilationWithFunctionPointersAndIl(source, il).VerifyDiagnostics();
```
>//CreateCompilationWithFunctionPointersAndIl(source, il).VerifyDiagnostics(); [](start = 12, length = 77)

Remove?

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -317,12 +801,14 @@ static void validateProperty(PropertySymbol property, Action<TypeSymbol> verifie
     public readonly delegate*<C, C> _field;
 }", symbolValidator: symbolValidator);
 
+            symbolValidator((ModuleSymbol)((Symbols.PublicModel.ModuleSymbol)verifier.Compilation.SourceModule).UnderlyingSymbol);
```
>(ModuleSymbol)((Symbols.PublicModel.ModuleSymbol)verifier.Compilation.SourceModule).UnderlyingSymbol [](start = 28, length = 100)

Consider extracting a helper method.

AlekseyTs commented:
>                 for (; paramIndex <= paramCount; paramIndex++)

This recovery doesn't feel appropriate in case of a function pointer. We should re-throw if we are decoding function pointer signature because we are very likely to attempt to decode something else after it, but there is no guarantee that the ```signatureReader``` is actually positioned after the function pointer. 

---
Refers to: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:1928 in 6d7c930. [](commit_id = 6d7c9305b094adb1e0dc1972c8902b8530ded376, deletion_comment = False)

AlekseyTs commented:
>                 for (; paramIndex <= paramCount; paramIndex++)

It should be fine to address that in a separate PR, but I think some tests added in this PR will be significantly affected.

---
In reply to: [618047228](https://github.com/dotnet/roslyn/pull/43355#issuecomment-618047228) [](ancestors = 618047228)

---
Refers to: src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:1928 in 6d7c930. [](commit_id = 6d7c9305b094adb1e0dc1972c8902b8530ded376, deletion_comment = False)

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -4261,6 +4892,226 @@ .maxstack  1
+    static void M(delegate*<void> func_1a, delegate*<string, void> func_1b, delegate*<int> func_2, int* int_1, int* int_2)
+    {
+        var compareResults = DoCompare(func_1a, func_1b);
+        Console.WriteLine(""func_1a == func_1b: "" + compareResults.Item1);
```
>compareResults.Item1 [](start = 53, length = 20)

Just curious: Why are the operators for the first pair evaluated with `DoCompare()` and the remaining pairs evaluated inline?

333fred replied:
I didn't think it was necessary or useful to verify the exact IL for each type of binary operator conversion, so I split out the operators for just the function pointer to function pointer operators and verified this one function. I left a comment for that a bit lower, but I probably should have put it here.

---
In reply to: [413345822](https://github.com/dotnet/roslyn/pull/43355#discussion_r413345822) [](ancestors = 413345822)

333fred replied:
Actually, I think it was here, but subsequent test additions moved it.

---
In reply to: [413346707](https://github.com/dotnet/roslyn/pull/43355#discussion_r413346707) [](ancestors = 413346707,413345822)

AlekseyTs commented:
Done with review pass (iteration 8). If you'd like to push some of the changes to a future PR, let's sync up offline.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/SymbolFactory.vb:
```diff
@@ -52,7 +52,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+
+        Friend Overrides Function IsAcceptedOutAttributeModifierType(type As TypeSymbol) As Boolean
+            ' VB doesn't deal with ref-readonly parameters, function pointers, or ref-return-types.
             Return False
```
>Return False [](start = 12, length = 12)

Is this code reachable? Consider adding a test hitting it.

333fred commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -1704,6 +1704,9 @@ private bool IsRef(BoundExpression receiver)
                     return ((BoundCall)receiver).Method.RefKind != RefKind.None;
 
+                case BoundKind.FunctionPointerInvocation:
+                    return ((BoundFunctionPointerInvocation)receiver).FunctionPointer.Signature.RefKind != RefKind.None;
```
This change is tested by `CodeGenFunctionPointersTests.FunctionPointerReturnTypeConstrainedCallVirtIfNotRef`. It's used by `EmitCall` when determining whether the receiver needs to be a constrained or boxed for a call virt.

333fred commented:
@AlekseyTs @cston addressed feedback.

---
In reply to: [618065936](https://github.com/dotnet/roslyn/pull/43355#issuecomment-618065936) [](ancestors = 618065936)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -2471,12 +2487,15 @@ internal bool DoesSignatureMatchEvent(TypeSymbol eventType, ParamInfo<TypeSymbol
+            System_Runtime_CompilerServices_Volatile = 1,
+            System_Runtime_InteropServices_InAttribute = 1 << 1,
+            System_Runtime_InteropServices_UnmanagedType = 1 << 2,
+            System_Runtime_CompilerServices_IsExternalInit = 1 << 3,
```
>System_Runtime_CompilerServices_IsExternalInit = 1 << 3, [](start = 12, length = 56)

Perhaps just leave a gap for this?

333fred replied:
I think if I left a gap it wouldn't merge cleanly, so I just copy/pasted to hopefully make the merge clean.

---
In reply to: [413415975](https://github.com/dotnet/roslyn/pull/43355#discussion_r413415975) [](ancestors = 413415975)

AlekseyTs replied:
> I think if I left a gap it wouldn't merge cleanly, so I just copy/pasted to hopefully make the merge clean.

I think it will still require minor manual intervention given the addition of a line below, pretty much the same level of intervention as for a gap. But I might be wrong.

---
In reply to: [413958941](https://github.com/dotnet/roslyn/pull/43355#discussion_r413958941) [](ancestors = 413958941,413415975)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs:
```diff
@@ -1187,21 +1195,28 @@ internal ImmutableArray<LocalInfo<TypeSymbol>> GetLocalInfo(StandaloneSignatureH
                 info.CustomModifiers = DecodeModifiersOrThrow(ref signatureReader, AllowedRequiredModifierType.None, out typeCode, out _);
             }
-            else if (inAttributeFound)
+            else if (foundRequiredTypes != 0)
```
>0 [](start = 43, length = 1)

AllowedRequiredModifierType.None ?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -434,14 +851,65 @@ .maxstack  8
+
+            var comp = CreateCompilationWithFunctionPointersAndIl(source, ilSource);
+            comp.VerifyDiagnostics(
+                // (6,11): error CS8795: The calling convention of 'delegate*<void>' is not supported the language.
```
> the [](start = 101, length = 4)

Minor thing - missing "by"

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -434,14 +851,65 @@ .maxstack  8
+                (RefKind.None, IsVoidType()));
+
+            Assert.True(type.HasUseSiteError);
+            Assert.True(type.Signature.IsVararg);
```
>Assert.True(type.Signature.IsVararg); [](start = 12, length = 37)

Is there any negative tests for this property. I.e. we expect false and assert it is false.

333fred replied:
Will add to `CommonVerifyFunctionPointer`

---
In reply to: [413424009](https://github.com/dotnet/roslyn/pull/43355#discussion_r413424009) [](ancestors = 413424009)

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -4261,43 +4822,256 @@ .maxstack  1
-                arrayTypeVerifier(((ArrayTypeSymbol)typeSymbol).ElementType);
-            };
+        [Fact]
+        public void FunctionPointerReturnTypeConstrainedCallVirtIfNotRef()
```
>IfNotRef [](start = 64, length = 8)

Should this be "IfRef" instead?

jaredpar commented:
Integration failures are same as in the CI builds. Merging.
